### DISCUSSION TITLE
Protocols and ObservableObject: "Only concrete types..."

### DISCUSSION THREAD

**Discussion by jonduenas** [hmlongco/Factory#93] [Q&A] [answered]
Here's what I'm trying to do...
```swift
protocol MyProtocol: ObservableObject {}

class MyClass: MyProtocol {}

extension Container {
    var myClass: Factory<any MyProtocol> {
        self { MyClass() }
    }
}

struct MyView: View {
    @InjectedObject(\.myClass) private var myClass // Error here
    
    // ...
}
``` 

The error I get is `Type 'any MyProtocol' cannot conform to 'ObservableObject'` and `Only concrete types such as structs, enums and classes can conform to protocols`.

I think I understand why this is happening. For example, if I did...

```swift
struct MyView<T: MyProtocol>: View {
    @ObservedObject private var myClass: T
    
    // ...
}
```
That compiles. But, I'm struggling to understand how to get this to work with Factory. Is the only way to make this work with Factory is to have the `ObservableObject` be a concrete type?

**Comment by hmlongco** [comment, marked as answer]
You might take a look at the following.

https://medium.com/swlh/did-swiftui-just-kill-protocol-oriented-programming-c09c2787172

Making the VM a protocol is a bad idea, and suffers from a plethora of problems.

https://stackoverflow.com/questions/59503399/how-to-define-a-protocol-as-a-type-for-a-observedobject-property

This also leads into all sorts of fun issues trying to setup previews, and trying to bind to published values on the view model.

This is really an issue with Swift and when Apple, in their infinite wisdom, decided that ObservableObject should be based on associated types. There really isn't a lot I can do about it other than say, "Don't do that."

Seriously.

**Reply by jonduenas** [reply]
After I wrote out my question, I thought to myself, "I bet I'm gonna have to have the `ObservableObject` contain the protocol rather than conform to the protocol." So it's good to see you confirm that. I also hadn't considered this particular object a "view model", but changing my perspective and seeing this as essentially playing a similar role does provide the path towards the answer. Appreciate your reply (and all the work you do on Factory).

**Reply by hmlongco** [reply]
Thanks. It’s a very common scenario that I should address in the documentation. But it’s also why I wrote the article.
